### PR TITLE
Fixed #23242: Added approximate_date_hierarchy option

### PR DESCRIPTION
**Pull request by gavinwahl**
<https://code.djangoproject.com/ticket/23242>

### CODE REVIEW

**Comment by timgraham** [comment]
Maybe we could use an enum-like class to define the available choices for the option? "0, 1, 2, 3, True, or False" doesn't seem ideal due magic numbers and lack of readability at a glance.

**Comment by gavinwahl** [comment]
@timgraham I don't really care either way. I chose integers because a higher number number indicates more estimation. What would the names of the enum values be?

**Comment by timgraham** [comment]
I left some ideas on the django-developers thread.

**Comment by cabello** [comment]
This is looking promising I am suffering from the fact that date hierarchy performance suffers on big tables. :)

**Comment by gavinwahl** [comment]
Updated

**Comment by timgraham** [comment]
Looks mostly good. Could you please rebase your branch so it merges cleanly and update the versions to 1.10?

File: django/contrib/admin/templatetags/admin_list.py [hunk 1]
```diff
@@ -390,7 +413,15 @@ def date_hierarchy(cl):
                 } for month in months]
             }
         else:
-            years = getattr(cl.queryset, dates_or_datetimes)(field_name, 'year')
+            if approximate >= ApproximateWith.YEARS:
```
Comment (charettes) [inline django/contrib/admin/templatetags/admin_list.py]: Using `elif approximate >= ApproximateWith.YEARS:` and keeping the final `else` clause as is would remove a nesting level.

**Comment by gavinwahl** [comment]
Really disappointing that this got left out of 1.9, considering it was ready before the cutoff date. I don't know if I'll be able to get time to do the rebase -- I've already exhausted all the time I had available for it, and my client probably won't even be able to benefit from it.

**Comment by timgraham** [comment]
Sorry, you updated it the day after the feature freeze.

**Pull request closed without merging**